spec/tmx_document_spec.rb

```diff
@@ -0,0 +1,97 @@
+# encoding: UTF-8
+
+require 'spec_helper'
+
+include Rosette::Integrations
+
+describe SmartlingIntegration::TmxDocument do
+  let(:locale_code) { 'de-DE' }
+  let(:result_hash) { {} }
+
+  let(:document) do
+    SmartlingIntegration::TmxDocument.new(locale_code) do |meta_key, key|
+      result_hash[meta_key] = key
+    end
+  end
+
+  let(:parser) do
+    Nokogiri::XML::SAX::Parser.new(document)
+  end
+
+  it 'identifies keys and meta keys' do
+    tmx_contents = %Q{
+      <tmx version="1.4">
+        <body>
+          <tu tuid="79b371014a8382a3b6efb86ec6ea97d9" segtype="block">
+            <prop type="x-segment-id">0</prop>
+            <prop type="smartling_string_variant">six.hours</prop>
+            <tuv xml:lang="en-US"><seg>6 hours</seg></tuv>
+            <tuv xml:lang="de-DE"><seg>6 Stunden</seg></tuv>
+          </tu>
+        </body>
+      </tmx>
+    }
+
+    parser.parse(tmx_contents)
+    expect(result_hash).to eq(
+      'six.hours' => '6 Stunden'
+    )
+  end
+
+  it 'falls back to the tuid if no meta key exists' do
+    tmx_contents = %Q{
+      <tmx version="1.4">
+        <body>
+          <tu tuid="79b371014a8382a3b6efb86ec6ea97d9" segtype="block">
+            <prop type="x-segment-id">0</prop>
+            <tuv xml:lang="en-US"><seg>6 hours</seg></tuv>
+            <tuv xml:lang="de-DE"><seg>6 Stunden</seg></tuv>
+          </tu>
+        </body>
+      </tmx>
+    }
+
+    parser.parse(tmx_contents)
+    expect(result_hash).to eq(
+      '79b371014a8382a3b6efb86ec6ea97d9' => '6 Stunden'
+    )
+  end
+
+  it 'does not return strings that have a segment id greater than zero' do
+    tmx_contents = %Q{
+      <tmx version="1.4">
+        <body>
+          <tu tuid="79b371014a8382a3b6efb86ec6ea97d9" segtype="block">
+            <prop type="x-segment-id">1</prop>
+            <tuv xml:lang="en-US"><seg>6 hours</seg></tuv>
+            <tuv xml:lang="de-DE"><seg>6 Stunden</seg></tuv>
+          </tu>
+        </body>
+      </tmx>
+    }
+
+    parser.parse(tmx_contents)
+    expect(result_hash).to eq({})
+  end
+
+  context 'with a different locale' do
+    let(:locale_code) { 'fr-FR' }
+
+    it 'does not return strings in the wrong locale' do
+      tmx_contents = %Q{
+        <tmx version="1.4">
+          <body>
+            <tu tuid="79b371014a8382a3b6efb86ec6ea97d9" segtype="block">
+              <prop type="x-segment-id">0</prop>
+              <tuv xml:lang="en-US"><seg>6 hours</seg></tuv>
+              <tuv xml:lang="de-DE"><seg>6 Stunden</seg></tuv>
+            </tu>
+          </body>
+        </tmx>
+      }
+
+      parser.parse(tmx_contents)
+      expect(result_hash).to eq({})
```

Comment:
I was thinking it should return a blank set, since you always download and parse a .tmx file in the context of a locale. This behavior is mostly incidental however because .tmx files have entries for both "en-US" and the target locale (in this case, "de-DE"), and the code tries to ignore the "en-US" locale. It makes sense to me for this method to throw an error though because there should always be an entry for the locale in each `<tu>` element. If not, then something obviously went wrong on the Smartling side. Rather than pull pull potentially incorrect data, we should probably halt the pulling process.